webapp/src/main/java/org/opentestsystem/rdw/admin/service/impl/DefaultCsvValidationService.java

```diff
@@ -0,0 +1,200 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.google.common.collect.Sets.newHashSet;
+import static org.apache.commons.lang.StringUtils.isBlank;
+
+/**
+ * Default implementation of a CsvValidationService
+ */
+@Service
+public class DefaultCsvValidationService implements CsvValidationService {
+
+    private static final String ValidationFormat = "Row: %1$d Failure: %2$s\n";
+    private static final String MoreFailuresFormat = "With %1$d additional validation failures\n";
+    private static final int MaxFailureMessages = 100;
+
+    private final SchoolRepository schoolRepository;
+
+    @Autowired
+    public DefaultCsvValidationService(final SchoolRepository schoolRepository) {
+        this.schoolRepository = schoolRepository;
+    }
+
+    @Override
+    public List<CsvValidationFailure> validateHeaders(final Set<String> csvHeaders) {
+        final List<CsvValidationFailure> failures = new ArrayList<>();
+        final Set<String> headers = csvHeaders.stream().map(String::toLowerCase).collect(Collectors.toSet());
+        if (!headers.containsAll(RequiredHeaders)) {
+            final Set<String> missing = newHashSet(RequiredHeaders);
+            missing.removeAll(headers);
+            failures.add(CsvValidationFailure.builder()
+                    .row(0)
+                    .message("Missing headers " + missing + ", found " + headers)
+                    .build());
+        }
+        return failures;
+    }
+
+    @Override
+    public List<CsvValidationFailure> validateRecords(final Iterator<CSVRecord> recordIterator,
+                                                      final PermissionScope permissionScope) {
+        final List<CsvValidationFailure> validationFailures = new ArrayList<>();
+
+        final Set<String> availableSchools = schoolRepository.findAll(permissionScope).stream()
+                .map(School::getNaturalId)
+                .collect(Collectors.toSet());
+
+        final Set<String> completedSchools = new HashSet<>();
+        final Set<String> completedSchoolGroupYears = new HashSet<>();
+
+        String currentSchool = null;
+        String currentSchoolGroupYear = null;
+        CsvSubject currentSchoolGroupYearSubject = null;
+
+        while (recordIterator.hasNext()) {
+            final CSVRecord record = recordIterator.next();
+            final String school = record.get(HeaderSchool);
+            final String group = record.get(HeaderGroup);
+            final Integer schoolYear = getAsInt(record, HeaderYear);
+            final CsvSubject subject = CsvSubject.valueOfCaseInsensitive(record.get(HeaderSubject));
+
+            if (isBlank(school) && isBlank(group) && schoolYear == null) {
+                //Not a viable record, skip
```

Comment:
I meant to call this out.  Does this make sense?  Or should I just pull this check out?  I think we're set to *not* ignore empty lines, but *should* we ignore empty lines?